himalaya: 0.6.0 -> 0.7.1

soywod opened the pull request:
###### Description of changes

Major version bump, see [CHANGELOG](https://github.com/soywod/himalaya/releases/tag/v0.7.0). I am the author of the tool, and I feel confortable enough now with Nix to be part of maintainers in order to reduce delay between releases.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

cc @toastal @yanganto

soywod committed:
prepare v0.6.2

soywod committed:
bump v0.7.0

soywod committed:
Merge branch 'NixOS:master' into master

toastal commented:
@soywod Will this and should this cover the Vim plugin (in a separate commit)? If the Vim is doing a automated update, these could get out of sync.

toastal commented:
Of note: `nix-shell -p nixpkgs-review --run "nixpkgs-review pr 215330"` did succeed for building on my machine as well.

yanganto commented on pkgs/applications/networking/mailreaders/himalaya/default.nix:
```diff
@@ -34,12 +36,17 @@ rustPlatform.buildRustPackage rec {
-  # flag added because without end-to-end testing is ran which requires
-  # additional tooling and servers to test
+  # TODO: is there a way to spawn an IMAP server to run integration
+  # tests?
```
Hi @soywod,
Nice to meet you here. :wave: 

It could be good to cover the integration test for the app in Rust.
Rust By Example also suggests a way to set up an integration test.
https://doc.rust-lang.org/rust-by-example/testing/integration_testing.html

If you think, the integration test will not really run every time and need to run dependent on the environment smartly without flags.  `test-with` could be a solution for you.
https://github.com/yanganto/test-with

Here is my suggestion to do this, and happy to help you set up the integration test.
An IMAP server executable binary could be set in `nativeBuildInputs` for nix,  such that the test in Rust can now setup it up and do integration tests.  And use [executable condition](https://github.com/yanganto/test-with#executable-condition) of `test-with`, such that we don't need a cargoTestFlag, and other contributors will not worry to run test cases with the corresponding IMAP server on their environment.

toastal replied:
`nativeBuildInputs` or `nativeCheckInputs`??

yanganto replied:
@toastal Good catch,  thanks.

yanganto replied:
Hi @soywod  
Here is some information could be helpful.

`openjdk17-bootstrap` and `greenmail` work well with this command 

`java -Dgreenmail.setup.test.all -Dgreenmail.hostname=0.0.0.0 -Dgreenmail.auth.disabled -Dgreenmail.verbose -jar /tmp/greenmail-standalone-1.6.13.jar`

soywod replied:
@yanganto nice to meet you back there :slightly_smiling_face: I already have integration tests working for the tool, and I also know how to launch them even in a CI (I mostly use this command: `docker run -it --rm -p 3025:3025 -p 3110:3110 -p 3143:3143 -p 3465:3465 -p 3993:3993 -p 3995:3995 -e GREENMAIL_OPTS='-Dgreenmail.setup.test.all -Dgreenmail.hostname=0.0.0.0 -Dgreenmail.auth.disabled -Dgreenmail.verbose' greenmail/standalone:1.6.11`). My question was more: how to start this server during nix tests?

> An IMAP server executable binary could be set in `nativeCheckInputs` for nix (It is a dependency for `cargo test`), such that the test in Rust can now setup it up and do integration tests when the nix package is building.
> […]
> `openjdk17-bootstrap` and `greenmail` work well with this command

Sounds great! But I did not understand *when* to launch the IMAP server?

soywod replied:
Thank you for all those details, but I still don't get *how* and *when* to spawn the `.jar` :sweat_smile: the simplest for everyone would be to use Rust crates instead of an external `.jar`, this way you control when to start / stop it. The best project I found so far is [Stalwart](https://stalw.art/), but they expose only a CLI not the lib (I just asked [the question](https://github.com/stalwartlabs/jmap-server/discussions/31), let's see). Anyway sounds a bit complicated for now, maybe we could leave it for a next PR?

yanganto replied:
These might be a clear response for you.

```text
When -> the check phase of the flow to build the nix package, but handled by `buildRustPackage` for Rust app.
How  ->  1. need to customize `buildRustPackage` and 
         2. docker image(not reasonable) or `.jar` should be included in this commit of nixpkgs.  
  or
  
How  ->  1. control the integration test flow in Rust test case,
         2. include `.jar` in Rust project or the commit here
         3. add `nativeCheckInputs = [openjdk17-bootstrap];`
```

Currently, this PR is good, and this thread is just a comment/suggestion, not a required review.  :smile: And I am not sure it is also good for @toastal.

I have no permission to merge this, the thing I only have is to reject and close this PR.
Disabled some tests not work with Nix flow in the current PR looks good to me,
so I just provided the information about how to make a compatible test flow to make the integration test works with NixOS. 

Nix is a robust system, Rust is a robust language, and we all love these.:heart:
Somehow we need a new thinking for these, using a docker image to test an app is not suitable, there is nowhere to put `docker run ...` and make everything works.  I hope you can understand, I just want to use my experience and try to save you time without pain.  Thanks for your time to read a lot of words :pray:

soywod replied:
Thank you again for your replies, I understand better now. I just created [an issue](https://todo.sr.ht/~soywod/pimalaya/36) to run integration tests without docker images, which will fix issues for everyone (nix included).

soywod commented:
> @soywod Will this and should this cover the Vim plugin (in a separate commit)? If the Vim is doing a automated update, these could get out of sync.

I totally forgot about the Vim plugin, I think it is better to do it within this PR. I will take care of it!

toastal commented:
> I think it is better to do it within this PR

Yeah. They are coupled and there is often a lag between review approval and merging so it'd be better to get them ushered in as one.

soywod commented:
> @soywod Will this and should this cover the Vim plugin (in a separate commit)? If the Vim is doing a automated update, these could get out of sync.

Wait, in fact it should already work? Because the vim plugin comes with a [flake](https://git.sr.ht/~soywod/himalaya-vim/tree/master/item/flake.nix#L18) that take the `himalaya` package from the `develop` branch (though, I'm not sure if it is the best approach, maybe it is safer to take it from `pkgs.himalaya`.

soywod committed:
himalaya: adjust todo comment about cargo test flags

toastal commented:
@soywod See `pkgs/applications/editors/vim/plugins/`. I believe the route to go would be to: remove the `https://git.sr.ht/~soywod/himalaya-vim,,` plugin from `vim-plugin-names`, and move it to manually bump via `overrides.nix`. (an example would be `LanguageClient-neovim`).

soywod committed:
himalaya: bump v0.7.1

soywod commented:
@toastal I just pushed, is it what you had in mind? I also bumped the version with the latest release v0.7.1.

toastal commented on pkgs/applications/editors/vim/plugins/vim-plugin-names:
```diff
@@ -1175,3 +1174,4 @@ https://github.com/ziglang/zig.vim/,,
 https://github.com/mickael-menu/zk-nvim/,HEAD,
 https://github.com/troydm/zoomwintab.vim/,,
 https://github.com/nanotee/zoxide.vim/,,
+https://git.sr.ht/~soywod/himalaya-vim,,
```
I don't think you want this. If you update your Vim plugin for a major release and it's not merged yet to `nixpkgs`'s default branch (happens a lot, even if approved and ready), then downstream users will get into a broken state as the Vim plugins get updated on a regular schedule. You could either tack on a branch like `v1.x` that it's safe to pull updates from and not break, or it is manually updated in the `overrides.nix` in the Vim plugins folder.

toastal replied:
For context, I originally added it in here _because_ getting feedback and a merge took too long. It'd be a good idea to leave a note/comment in the `default.nix` though to double check the Vim plugin version or something.

soywod replied:
I see. I will propose sth and let you know!

soywod committed:
make himalaya-vim derivation use versions to reduce breaking changes

soywod commented:
@toastal I pushed sth, let me know. Basically:

- I updated the `flake.nix` of the himalaya vim plugin to use the himalaya CLI derivation from `develop` and to take its sources from `self` (the local directory). I also apply git flags the same way I do for the CLI, this way each CLI version matches a same vim version.
- I updated `override.nix` to override the himalaya package (take the one from nixpkgs itself instead of `develop`) and to override the sources (taking the archive matching the version of this same himalaya package)

soywod commented on pkgs/applications/editors/vim/plugins/overrides.nix:
```diff
@@ -578,28 +580,30 @@ self: super: {
-    postInstall = ''
-      ln -s ${nodeDep}/lib/node_modules/markdown-preview/node_modules $out/app
-    '';
+  markdown-preview-nvim = super.markdown-preview-nvim.overrideAttrs (old:
```
Starting from this line, changes come from `nixpkgs-fmt`.

toastal replied:
:| They'll likely ask you not to run `nixpkgs-fmt` on the file even though a _lot_ of the code *does* use that formatter and/or style (though I'm sure there's a lot of us that would prefer the whole repo used a single formatter for consistency). You could reset the commit and try `git add -p` and just patch in your lines.

soywod replied:
Changes reverted!

toastal commented on pkgs/applications/editors/vim/plugins/overrides.nix:
```diff
@@ -488,12 +489,13 @@ self: super: {
-      substituteInPlace autoload/himalaya/request.vim \
-        --replace "'himalaya" "'${himalaya}/bin/himalaya"
-    '';
+    buildInputs = [ himalaya ];
```
This all looks good though. Great to see we don't have to patch the package to get those boolean checks now!

soywod committed:
himalaya: revert formatting

edmundmiller reviewed (approved):
Awesome! Came here from trying to bump the version. Glad to see you maintaining it @soywod !

toastal commented on pkgs/applications/networking/mailreaders/himalaya/default.nix:
```diff
@@ -34,12 +36,17 @@ rustPlatform.buildRustPackage rec {
-  # flag added because without end-to-end testing is ran which requires
-  # additional tooling and servers to test
+  # TODO: remove this flag once this issue is fixed:
+  # https://todo.sr.ht/~soywod/pimalaya/36
```
Is this `pimalaya` a typo though? 🙃

soywod replied:
In fact `himalaya` becomes a subproject of a bigger one called `pimalaya`. The aim of this bigger project is to gather tools working the same way as `himalaya` (Rust lib + interfaces) in order to improve the way we manage our [Personal Information](https://en.wikipedia.org/wiki/Personal_information_manager). It changes nothing for `himalaya` itself, just the [website](https://pimalaya.org/), the [bug tracker](https://todo.sr.ht/~soywod/pimalaya) and the [mailing list](https://lists.sr.ht/~soywod/pimalaya) use the name `pimalaya` :slightly_smiling_face:

toastal replied:
I wasn't 100% sure on this one, but cool.

michaeladler commented:
@soywod It would be nice if you could enable the notmuch-backend feature

soywod committed:
himalaya: add args to manage cargo features

yanganto reviewed (changes requested):
The `test_imap_backend` is moved into pimalaya(current in a different repo) not in the current himalaya repo anymore, this package runs the tests of `himalaya` not `pimalaya`.  Is this flag still required?

`cargoTestFlags = [ "--lib" ];`

soywod commented:
> The imap backend test is moved into pimalaya(in a split lib) not in
> the current himalaya repo anymore, this flag is not required now.

You right, we do not need this flag anymore! But it also means we do not
fully test Himalaya, only the CLI part. Maybe later in another we could
see how to run lib tests.

> By the way, there is a commit help you set up imap test server with
> `.jar` file in Rust, and the `bob.key` absent in the repo, so I can
> not help you complete the imap testcase.

Awesome, thank you! I will check it out whenever I can.

soywod committed:
remove useless cargo test flags

yanganto commented on pkgs/applications/networking/mailreaders/himalaya/default.nix:
```diff
@@ -34,12 +36,17 @@ rustPlatform.buildRustPackage rec {
-  # additional tooling and servers to test
+  # TODO: remove this flag once this issue is fixed:
+  # https://todo.sr.ht/~soywod/pimalaya/36
   cargoTestFlags = [ "--lib" ];
```
> By the way, there is a commit help you set up imap test server with `.jar` file in Rust, and the `bob.key` absent in the repo, so I can not help you complete the imap testcase. [yanganto/himalaya-lib@fd2f417](https://github.com/yanganto/himalaya-lib/commit/fd2f417c02df9a6f52fa7fc94c89016898f17824)

Hi @soywod,
I got your email, and here is the commit to help you remove the `.jar`.
Is this patch your mentioned in email?
